pkg/controller/vsphere/actuator.go

```diff
@@ -71,25 +75,34 @@ func (a *Actuator) Create(ctx context.Context, machine *machinev1.Machine) error
 		fmtErr := fmt.Errorf(scopeFailFmt, machine.GetName(), err)
 		return a.handleMachineError(machine, fmtErr, createEventAction)
 	}
-	if err := newReconciler(scope).create(); err != nil {
-		if err := scope.PatchMachine(); err != nil {
-			return err
+
+	// Ensure we're not reconciling a stale machine by checking our task-id.
+	// This is a workaround for a cache race condition.
+	if val, ok := a.TaskIDCache[machine.Name]; ok {
+		if val != scope.providerStatus.TaskRef {
+			klog.Errorf("%s: machine object missing expected provider task ID, requeue", machine.GetName())
+			return &machinecontroller.RequeueAfterError{RequeueAfter: requeueAfterSeconds * time.Second}
 		}
+	}
+
+	var retErr error
+	err = newReconciler(scope).create()
+	// save the taskRef in our cache in case of any error with patch.
+	if scope.providerStatus.TaskRef != "" {
+		a.TaskIDCache[machine.Name] = scope.providerStatus.TaskRef
+	}
+	if err != nil {
 		fmtErr := fmt.Errorf(reconcilerFailFmt, machine.GetName(), createEventAction, err)
-		return a.handleMachineError(machine, fmtErr, createEventAction)
+		retErr = a.handleMachineError(machine, fmtErr, createEventAction)
```

Comment:
Nit: Maybe add a comment here to explain why we aren't returning here immediately? WDYT?